samcli/lib/iac/plugins_interfaces.py

```diff
@@ -0,0 +1,902 @@
+        self._lookup_path_type = lookup_path_type
+
+
+class ProjectFindRule:
+    @abc.abstractmethod
+    @property
+    def file_rules(self) -> List[str]:
+        """
+        a list of regex to filter the file names or suffixes
+        """
+
+    @abc.abstractmethod
+    @property
+    def content_rules(self) -> Dict[str, str]:
+        """
+        regex to filter the file context. Key is the regex of filename
+        Value is file content filter with regex.
+        """
+
+
+class PluginContext:
+    def __init__(self, context: Dict[str, Any]):
+        self._context = context
+
+    @property
+    def context_map(self):
+        """
+        the context retrieved from command line, its key is the command line option
+        name, value is corresponding input
+        """
+        return self._context
+
+
+class IaCPluginInterface(metaclass=abc.ABCMeta):
+    """
+    Interface for an IaC Plugin
+    We only require two methods here - get_project and write_project
+    """
+
+    def __init__(self, context: PluginContext):
+        self._context = context
+
+    @abc.abstractmethod
+    def read_project(self, lookup_paths: List[str]) -> SamCliProject:
+        """
+        Read and parse template of that IaC Platform
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def write_project(self, project: SamCliProject, build_dir: str) -> bool:
+        """
+        Write project to a template (or a set of templates),
+        move the template(s) to build_path
+        return true if it's successful
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def update_packaged_locations(self, stack: Stack) -> bool:
+        """
+        update the locations of assets inside a stack after sam packaging
+        return true if it's successful
+        """
+        raise NotImplementedError
+
+
+class IaCPluginDefinition:
+    def __init__(
+        self,
+        plugin_name: str,
+        plugin_creator: Callable[[PluginContext], IaCPluginInterface],
+        plugin_detector: ProjectFindRule,
+        additional_options: List[Any],
+    ):
+        self._plugin_name = plugin_name
+        self._plugin_creator = plugin_creator
+        self._project_detector = plugin_detector
+        self._additional_options = additional_options
+
+    def create_plugin(self, context: PluginContext) -> IaCPluginInterface:
+        """
+        factory method to use plugin_creator function to instantiate an IaCPluginInterface object.
+        """
+        return self._plugin_creator(context)
+
+    @abc.abstractmethod
+    def project_detected(self, path: str) -> bool:
+        """
+        use the defined project_detector function to detect the type of a project
+        return true if a valid project is detected
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def valid_context(self, options: List[Any]) -> bool:
```

Comment:
Is `List[Any]` the correct type for click options?

Reply:
It's supposed to be map of option-value pair, but leave for extra valiations.

Reply:
`Dict[Any, Any]`?

Reply:
Agree with that.